modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -360,32 +368,76 @@ public void deleteBlobsIgnoringIfNotExists(OperationPurpose purpose, Iterator<St
         }
     }
 
+    /**
+     * Delete one partition of a batch of blobs
+     *
+     * @param purpose The {@link OperationPurpose} of the deletion
+     * @param clientReferenceHolder A holder for a {@link AmazonS3Reference}. Will be populated if empty, must be closed by the caller.
+     * @param partition The list of blobs to delete
+     * @param aex A holder for any exception(s) thrown during the deletion
+     */
     private void deletePartition(
         OperationPurpose purpose,
-        AmazonS3Reference clientReference,
+        AtomicReference<AmazonS3Reference> clientReferenceHolder,
         List<String> partition,
         AtomicReference<Exception> aex
     ) {
```

Comment:
I don't love this change in the API, but it seemed the lightest-touch way to allow the client reference to be re-acquired when a batch is retried while still allowing re-use between batches in the happy case. Open to moving to reference-per-batch if we think the extra allocations are fine.